### DISCUSSION TITLE
cFS Certification (NPR 7150.2C)

### DISCUSSION THREAD

**Discussion by pmehradnia** [nasa/cFS#452] [Q&A] [answered]
I am looking to get information on how cFS artefacts are classified in terms of NPR 7150.2C. Would the artefacts (cFE, OSAL, etc.) be individually certified, and would it be according to NPR 7150.2C Class A or Class E?

Thank you

**Comment by astrogeco** [comment, marked as answer]
The open-source releases are all classified as Class E and, under the [Apache 2.0 license](https://github.com/nasa/cFE/blob/main/LICENSE), 
> You are solely responsible for determining the
      appropriateness of using or redistributing the Work and assume any
      risks associated with Your exercise of permissions under this License.

We do provide the unit tests and some functional tests that can help you bump up the classification with some work on your end. If you want more hands-on support from our team you can contact cfs-program@lists.nasa.gov

**Reply by skliper** [reply]
Typically users will assemble a "distribution" that contains everything they need to meet their requirements, and certify the system as a whole relative to the NPR 7150.2C software classification of the final product.  As @astrogeco mentioned above, the open source repositories provide some of what is needed from the software aspect but full compliance is the responsibility of the user.  Note "certification" is not typically considered transferable due to the highly configurable nature of the system as a whole, the various elements, and variety of contexts it can be used in.  Many of the open source elements have been certified to B or A but not in a general purpose sort of way (not every option/configuration or possible combination has been certified).  The open source products are provided as "general purpose software" (Class E) as a basic framework that users can assemble and augment as needed.

**Reply by pmehradnia** [reply]
I understand that the Open Source cFS is certified as "Class E"; is there a formal certification document published somewhere that formally captures this classification?

**Reply by skliper** [reply]
Compliance with the applicable NPR 7150.2 requirements is managed internally.  The Open Source software is delivered externally as Apache 2.0, and there is no official/formal certification or classification documentation that accompanies the open source release.  We report what we can (best effort) in terms of testing performed and processes followed (see related version description documents), but in the end it's all still Apache 2.0 so the burden is on the user to determine if the software meets their needs/requirements/standards.

**Comment by JudeLabs** [comment]
I did stumble onto this article. https://issuu.com/nasagsfc/docs/winter_2021_final_web_version/s/11673289

There is a section in it that says:
"Multiple NASA centers, industry partners, and international space agencies are developing Gateway. Goddard’s Exploration and Space Communications Projects Division’s technology and exploration offices and the Software Engineering Division are collaborating with NASA’s Johnson Space Center in Houston to certify cFS as Class A: suitable for human-rated vehicles."

I haven't seen anything on Class A certification in anything that I can find.  I do realize that the developer will have to go through a certification,  but a previous certification at some level will help with adoption.  Does anyone know of any status along these lines?

**Reply by astrogeco** [reply]
You can learn more about NASA's software processes in our [Software Engineering Handbook](https://swehb.nasa.gov/). For a definition of the different classifications and their requirements, see <https://swehb.nasa.gov/display/SWEHBVD/SWE-020+-+Software+Classification> 

Note that classifications are project and hardware specific. There have been implementations of cFS "distributions" that are rated as NASA SW Classes C and B and there are probably others at lower levels. But, like we wrote above, these have to be classified independently on specific target hardware under unique configurations. That is why we say that cFS is "certifiable"; each user has the responsibility to run their verification and validation processes to turn that into their needed certification.
